python/pyspark/sql/dataframe.py

```diff
@@ -1272,16 +1272,18 @@ def replace(self, to_replace, value, subset=None):
         """Returns a new :class:`DataFrame` replacing a value with another value.
         :func:`DataFrame.replace` and :func:`DataFrameNaFunctions.replace` are
         aliases of each other.
+        Values `to_replace` and `value` should be homogeneous. Mixed string and numeric
```

Comment:
@holdenk I am not convinced. I am intentionally trying to be a bit vague to avoid suggestion that we support only `int` -> `int`or `float` -> `float`. We cannot use abstract base class (`numbers.Real`?) because we depend on types, not interfaces.

But if you feel strongly about it I'll trust your judgment. I would probably write `TypeVar('T', str, Union[float, int])` but I am pretty sure it wouldn't be welcome :) and I hope the other PR will make the requirements explicit.